back-end/routes/recommendations.js

```diff
@@ -18,34 +18,99 @@ const getAccessToken = async (userId) => {
+    let totalValence = 0;
+    let totalEnergy = 0;
+    let totalInstrumentalness = 0;
+    let totalCount = 0;
+
+    features.forEach(feature => {
+        totalValence += feature.valence;
+        totalEnergy += feature.energy;
+        totalInstrumentalness += feature.instrumentalness;
+        totalCount++;
+    });
+
+    const valence = totalCount > 0 ? totalValence / totalCount : 0;
+    const energy = totalCount > 0 ? totalEnergy / totalCount : 0;
+    const instrumentalness = totalCount > 0 ? totalInstrumentalness / totalCount : 0;
+
+    return {
+        valence,
+        energy,
+        instrumentalness
+    };
+}
+
+async function fetchAndProcessAudioFeatures(songs, access_token) {
+    const features = await fetchAudioFeatures(songs, access_token);
+    const averages = calculateGlobalFeatureAverages(features);
+    const lastSongFeatures = features.length > 0 ? features[features.length - 1] : null;
+    return {
+        averages,
+        lastSongFeatures
+    };
+}
+
+function gradingAlgorithm(lastSongFeatures, recommendationAudioFeatures, averages) {
+    try {
+        const scores = new Map();
+        const moodWeight = 0.40;
+        const danceWeight = 0.19;
+        const mixWeight = 0.41;
+        recommendationAudioFeatures.forEach(feature => {
+            const recommendId = feature.uri.split(':').pop();
+            const tempoScore = 1 - Math.abs(lastSongFeatures.tempo - feature.tempo) / lastSongFeatures.tempo;
+            const valenceScore = 1 - Math.abs(averages.valence - feature.valence) / averages.valence;
+            const danceScore = feature.danceability;
+            const totalScore = (valenceScore * moodWeight) + (danceScore * danceWeight) + (tempoScore * mixWeight)
+            scores.set(recommendId, totalScore);
+        });
+
+        const sortedScores = Array.from(scores).sort((a, b) => b[1] - a[1]);
+
+        let count = 1;
+        sortedScores.forEach(([id, diff]) => {
+            count++;
+        });
+    } catch (error) {
+        console.error("Error in gradingAlgorithm:", error);
+    }
+}
+
 const fetchRecommendations = async (songs, access_token) => {
     const lastSongIds = songs.map(song => (song.uri).split(':').pop()).slice(-5);
-    const stringOfIds = lastSongIds.join(',');
+    const commaSeparatedIds = lastSongIds.join(',');
     try {
-        const response = await fetch(`https://api.spotify.com/v1/recommendations?limit=20&seed_tracks=${stringOfIds}`, {
+        const response = await fetch(`https://api.spotify.com/v1/recommendations?limit=40&seed_tracks=${commaSeparatedIds}`, {
             headers: {
                 'Authorization': `Bearer ${access_token}`
             }
         });
         const recommendData = await response.json();
         const recommendations = recommendData.tracks;
-        const recommendIds = recommendations.map(track => (track.uri).split(':').pop());
-        return recommendIds;
+        return recommendations;
```

Comment:
Similar to the comments on https://github.com/DiegoLandaeta03/SoundStitch/pull/13/ let's try to use more descriptive names such as `recommendations` → `tracksRecommendations`.